### PR TITLE
BUG: Fixes histogram monotonicity check for unsigned bin values

### PR DESCRIPTION
**Pull request by b-carter**
Add test and fix the monotonicity check to support unsigned bin values, see #9222.

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -782,7 +782,7 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
         bins = bin_edges
     else:
         bins = asarray(bins)
-        if (np.diff(bins) < 0).any():
+        if not np.all(bins[1:] > bins[:-1]):
```
Comment (jaimefrio) [inline numpy/lib/function_base.py]: I find checks without a negation easier to read, i.e. `if np.any(bins[1:] <= bins[:-1]):`

**Review by jaimefrio — APPROVED** [review]
I have suggested a nitpick, otherwise this LGTM. Thanks!

**Comment by seberg** [comment]
There are more diff's on the bins further down, do they not matter? What happens if bins is two dimensional? Just out of curiosity to make sure it is an error.

**Comment by b-carter** [comment]
@jaimefrio I agree, that seems cleaner and maybe more efficient too. I just pushed an update.

@seberg, I'm not sure what you mean by more diff's on the bins further down. Could you give an example? Also, I don't think this breaks anything for the 2D case. It seems that is caught by a ``ValueError: all the input arrays must have same number of dimensions`` from a call to ``sa.searchsorted(bins[-1], 'right')`` on line 796. Maybe there should be an explicit check somewhere else though?

**Comment by seberg** [comment]
Lines 815 and 822? I don't care what error happens, to be honest, there is likely even a test, if not we could add one. Just wanted to make sure it does not currently silently convert it into a 1D array and we broke it.

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -782,7 +782,7 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
         bins = bin_edges
     else:
         bins = asarray(bins)
-        if (np.diff(bins) < 0).any():
+        if np.any(bins[1:] <= bins[:-1]):
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: This isn't equivalent - previously we allowed repeated bins, but this code does not.

I'd prefer `np.any(bins[:-1] > bins[1:])` here, which reads as `current > next`

**Review by eric-wieser — CHANGES_REQUESTED** [review]
Behaviour change for repeated bins

**Comment by charris** [comment]
Put the backport milestone off to 1.13.2. Looks like this is close to a fix, be good to finish it up.

**Comment by b-carter** [comment]
Sorry for the delay, just updated the PR with the suggestion from @eric-wieser.

**Comment by charris** [comment]
@eric-wieser Are you OK with this?

**Comment by charris** [comment]
@eric-wieser OK to put this in?

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
Might be worth flipping the condition to `not all(bins[:-1] <= bins[1:])` to change the nan behaviour, but that's out of scope for this PR.

**Pull request merged**